metricflow/plan_conversion/node_processor.py

```diff
@@ -59,14 +61,111 @@ class MultiHopJoinCandidate:
+    """Container class for managing information about whether and how to do filter predicate pushdown.
 
-    Stores time constraint information for applying pre-join time filters.
+    The time_range_constraint property holds the time window for setting up a time range filter expression.
     """
 
-    time_range_constraint: Optional[TimeRangeConstraint]
+    _PREDICATE_METADATA_KEY = "is_predicate_property"
+
+    time_range_constraint: Optional[TimeRangeConstraint] = dataclasses.field(metadata={_PREDICATE_METADATA_KEY: True})
+    pushdown_state: PredicatePushdownState = PredicatePushdownState.FULLY_ENABLED
+
+    def __post_init__(self) -> None:
+        """Validation to ensure pushdown properties are configured correctly.
+
+        In particular, this asserts that cases where pushdown is disabled cannot leak pushdown operations via
+        outside property access - if pushdown is disabled, no further pushdown operations of any kind are allowed
+        on that particular code branch.
+        """
+        if self.pushdown_state is PredicatePushdownState.FULLY_ENABLED:
+            return
+
+        invalid_predicate_property_names = {
+            field.name for field in dataclasses.fields(self) if field.metadata.get(self._PREDICATE_METADATA_KEY)
+        }
+
+        if self.pushdown_state is PredicatePushdownState.ENABLED_FOR_TIME_RANGE_ONLY:
+            # We don't do validation for time range constraint configuration - having None set in this state is the
+            # equivalent of disabling predicate pushdown, but that time constraint value might be updated later,
+            # and so we do not block overrides to (or from) None to avoid meaningless bookkeeping at callsites.
+            # Also, we keep the magic string name Python uses hidden in here instead of expanding access to it.
+            invalid_predicate_property_names.remove("time_range_constraint")
+
+        instance_configuration = dataclasses.asdict(self)
+        invalid_disabled_properties = {
+            property_name: value
+            for property_name, value in instance_configuration.items()
+            if property_name in invalid_predicate_property_names and value is not None
+        }
+
+        assert not invalid_disabled_properties, (
+            "Invalid pushdown parameter configuration! Disabled pushdown parameters cannot have properties "
+            "set that may lead to improper access and use in other contexts, as that can lead to unintended "
+            "filtering operations in cases where these properties are accessed without appropriate checks against "
+            "pushdown configuration. This indicates that pushdown is configured for limited scope operations, but "
+            f"other predicate properties are set to non-None values.\nFull configuration: {instance_configuration}\n"
+            f"Invalid predicate properties: {invalid_disabled_properties}"
+        )
+
+    @property
+    def is_pushdown_enabled(self) -> bool:
```

Comment:
Might be helpful to call this `is_pushdown_enabled_for_time_range` or something more specific so that other devs don't mistake this as an indication that pushdown is FULLY enabled!

Reply:
That isn't exactly what this does, but to your point inverting it to is_pushdown_disabled might be better.

Reply:
I've made a ton of updates and now this is EXACTLY what that does. I also put in the disabled thing but I'll probably remove it. I think I should stop tinkering here and move ahead, though.